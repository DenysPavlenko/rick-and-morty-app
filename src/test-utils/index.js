import checkPropTypes from 'check-prop-types';

export const findByTestAtt = (wrapper, val) =>
  wrapper.find(`[data-test="${val}"]`);

export const checkProps = (component, expectedProps) => {
  const propsErr = checkPropTypes(
    component.propTypes,
    expectedProps,
    'props',
    component.name
  );
  return propsErr;
};
