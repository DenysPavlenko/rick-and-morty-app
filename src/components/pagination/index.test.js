/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount } from 'enzyme';
import { findByTestAtt } from 'test-utils';
import Pagination from './index';

const dummyProps = {
  className: '',
  pages: 34,
  page: 1,
  setPage: () => {},
};

const setup = (props = {}) => mount(<Pagination {...props} />);

describe('Pagination component', () => {
  test('renders without error', () => {
    const wrapper = setup({ ...dummyProps });
    const component = findByTestAtt(wrapper, 'pagination');
    expect(component.length).toBe(1);
  });
  test('Renders 6 buttons in a row if there are less then 7 pages', () => {
    const wrapper = setup({ ...dummyProps, pages: 6 });
    expect(global.toJson(wrapper)).toMatchSnapshot();
  });
  test('Renders 5 buttons and 2 dots if the active page is in the center', () => {
    const wrapper = setup({ ...dummyProps, page: 10 });
    expect(global.toJson(wrapper)).toMatchSnapshot();
  });
  test('Renders 5 buttons, dots and 1 button if there more then 6 pages', () => {
    const wrapper = setup({ ...dummyProps, pages: 7 });
    expect(global.toJson(wrapper)).toMatchSnapshot();
  });
  test('Renders first page, dots and 5 buttons if the active page is in the end', () => {
    const wrapper = setup({ ...dummyProps, page: 33 });
    expect(global.toJson(wrapper)).toMatchSnapshot();
  });

  test('Changes page to previous on left arrow click', () => {
    const mockSetPage = jest.fn();
    const wrapper = setup({ ...dummyProps, page: 2, setPage: mockSetPage });
    const component = findByTestAtt(wrapper, 'pagination-chevron-left');
    component.simulate('click');
    expect(mockSetPage.mock.calls.length).toBe(1);
  });
  test('Doesn"t change page on left arrow click if the page number equals to 1', () => {
    const mockSetPage = jest.fn();
    const wrapper = setup({ ...dummyProps, page: 1, setPage: mockSetPage });
    const component = findByTestAtt(wrapper, 'pagination-chevron-left');
    component.props().onClick();
    expect(mockSetPage.mock.calls.length).toBe(0);
  });

  test('Changes page on button click', () => {
    const mockSetPage = jest.fn();
    const wrapper = setup({ ...dummyProps, page: 3, setPage: mockSetPage });
    const component = findByTestAtt(wrapper, 'pagination-button').at(1);
    component.simulate('click');
    expect(mockSetPage.mock.calls.length).toBe(1);
  });
  test('Doesn"t change page on button click if page equals to activePage', () => {
    const mockSetPage = jest.fn();
    const wrapper = setup({ ...dummyProps, page: 3, setPage: mockSetPage });
    const component = findByTestAtt(wrapper, 'pagination-button').at(2);
    component.props().onClick();
    expect(mockSetPage.mock.calls.length).toBe(0);
  });

  test('Changes page to next on right arrow click', () => {
    const mockSetPage = jest.fn();
    const wrapper = setup({ ...dummyProps, page: 1, setPage: mockSetPage });
    const component = findByTestAtt(wrapper, 'pagination-chevron-right');
    component.simulate('click');
    expect(mockSetPage.mock.calls.length).toBe(1);
  });
  test('Doesn"t change page on right arrow click if the page number equals to pages number', () => {
    const mockSetPage = jest.fn();
    const wrapper = setup({ ...dummyProps, page: 34, setPage: mockSetPage });
    const component = findByTestAtt(wrapper, 'pagination-chevron-right');
    component.props().onClick();
    expect(mockSetPage.mock.calls.length).toBe(0);
  });
});
