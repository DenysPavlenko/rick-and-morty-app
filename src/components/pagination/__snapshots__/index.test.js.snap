// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component Renders 5 buttons and 2 dots if the active page is in the center 1`] = `
<Pagination
  className=""
  page={10}
  pages={34}
  setPage={[Function]}
>
  <nav
    className="pagination"
    data-test="pagination"
  >
    <ul
      className="pagination-list"
    >
      <li
        className="pagination-item"
      >
        <button
          className="pagination-button"
          data-test="pagination-chevron-left"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <ForwardRef(SvgChevronLeft)
            className="pagination-chevron"
          >
            <svg
              className="pagination-chevron"
            >
              chevron-left.svg
            </svg>
          </ForwardRef(SvgChevronLeft)>
        </button>
      </li>
      <li
        className="pagination-item"
        key="0"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          1
        </button>
      </li>
      <li
        className="pagination-item"
        key="1"
      >
        <button
          className="pagination-button pagination-button-disabled"
          data-test="pagination-dots"
          disabled={true}
          type="button"
        >
          ...
        </button>
      </li>
      <li
        className="pagination-item"
        key="2"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          9
        </button>
      </li>
      <li
        className="pagination-item"
        key="3"
      >
        <button
          className="pagination-button pagination-button-active"
          data-test="pagination-button"
          disabled="active"
          onClick={[Function]}
          type="button"
        >
          10
        </button>
      </li>
      <li
        className="pagination-item"
        key="4"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          11
        </button>
      </li>
      <li
        className="pagination-item"
        key="5"
      >
        <button
          className="pagination-button pagination-button-disabled"
          data-test="pagination-dots"
          disabled={true}
          type="button"
        >
          ...
        </button>
      </li>
      <li
        className="pagination-item"
        key="6"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          34
        </button>
      </li>
      <li
        className="pagination-item"
      >
        <button
          className="pagination-button"
          data-test="pagination-chevron-right"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <ForwardRef(SvgChevronRight)
            className="pagination-chevron"
          >
            <svg
              className="pagination-chevron"
            >
              chevron-right.svg
            </svg>
          </ForwardRef(SvgChevronRight)>
        </button>
      </li>
    </ul>
  </nav>
</Pagination>
`;

exports[`Pagination component Renders 5 buttons, dots and 1 button if there more then 6 pages 1`] = `
<Pagination
  className=""
  page={1}
  pages={7}
  setPage={[Function]}
>
  <nav
    className="pagination"
    data-test="pagination"
  >
    <ul
      className="pagination-list"
    >
      <li
        className="pagination-item"
      >
        <button
          className="pagination-button pagination-button-disabled"
          data-test="pagination-chevron-left"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <ForwardRef(SvgChevronLeft)
            className="pagination-chevron"
          >
            <svg
              className="pagination-chevron"
            >
              chevron-left.svg
            </svg>
          </ForwardRef(SvgChevronLeft)>
        </button>
      </li>
      <li
        className="pagination-item"
        key="0"
      >
        <button
          className="pagination-button pagination-button-active"
          data-test="pagination-button"
          disabled="active"
          onClick={[Function]}
          type="button"
        >
          1
        </button>
      </li>
      <li
        className="pagination-item"
        key="1"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          2
        </button>
      </li>
      <li
        className="pagination-item"
        key="2"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          3
        </button>
      </li>
      <li
        className="pagination-item"
        key="3"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          4
        </button>
      </li>
      <li
        className="pagination-item"
        key="4"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          5
        </button>
      </li>
      <li
        className="pagination-item"
        key="5"
      >
        <button
          className="pagination-button pagination-button-disabled"
          data-test="pagination-dots"
          disabled={true}
          type="button"
        >
          ...
        </button>
      </li>
      <li
        className="pagination-item"
        key="6"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          7
        </button>
      </li>
      <li
        className="pagination-item"
      >
        <button
          className="pagination-button"
          data-test="pagination-chevron-right"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <ForwardRef(SvgChevronRight)
            className="pagination-chevron"
          >
            <svg
              className="pagination-chevron"
            >
              chevron-right.svg
            </svg>
          </ForwardRef(SvgChevronRight)>
        </button>
      </li>
    </ul>
  </nav>
</Pagination>
`;

exports[`Pagination component Renders 6 buttons in a row if there are less then 7 pages 1`] = `
<Pagination
  className=""
  page={1}
  pages={6}
  setPage={[Function]}
>
  <nav
    className="pagination"
    data-test="pagination"
  >
    <ul
      className="pagination-list"
    >
      <li
        className="pagination-item"
      >
        <button
          className="pagination-button pagination-button-disabled"
          data-test="pagination-chevron-left"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <ForwardRef(SvgChevronLeft)
            className="pagination-chevron"
          >
            <svg
              className="pagination-chevron"
            >
              chevron-left.svg
            </svg>
          </ForwardRef(SvgChevronLeft)>
        </button>
      </li>
      <li
        className="pagination-item"
        key="0"
      >
        <button
          className="pagination-button pagination-button-active"
          data-test="pagination-button"
          disabled="active"
          onClick={[Function]}
          type="button"
        >
          1
        </button>
      </li>
      <li
        className="pagination-item"
        key="1"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          2
        </button>
      </li>
      <li
        className="pagination-item"
        key="2"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          3
        </button>
      </li>
      <li
        className="pagination-item"
        key="3"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          4
        </button>
      </li>
      <li
        className="pagination-item"
        key="4"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          5
        </button>
      </li>
      <li
        className="pagination-item"
        key="5"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          6
        </button>
      </li>
      <li
        className="pagination-item"
      >
        <button
          className="pagination-button"
          data-test="pagination-chevron-right"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <ForwardRef(SvgChevronRight)
            className="pagination-chevron"
          >
            <svg
              className="pagination-chevron"
            >
              chevron-right.svg
            </svg>
          </ForwardRef(SvgChevronRight)>
        </button>
      </li>
    </ul>
  </nav>
</Pagination>
`;

exports[`Pagination component Renders first page, dots and 5 buttons if the active page is in the end 1`] = `
<Pagination
  className=""
  page={33}
  pages={34}
  setPage={[Function]}
>
  <nav
    className="pagination"
    data-test="pagination"
  >
    <ul
      className="pagination-list"
    >
      <li
        className="pagination-item"
      >
        <button
          className="pagination-button"
          data-test="pagination-chevron-left"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <ForwardRef(SvgChevronLeft)
            className="pagination-chevron"
          >
            <svg
              className="pagination-chevron"
            >
              chevron-left.svg
            </svg>
          </ForwardRef(SvgChevronLeft)>
        </button>
      </li>
      <li
        className="pagination-item"
        key="0"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          1
        </button>
      </li>
      <li
        className="pagination-item"
        key="1"
      >
        <button
          className="pagination-button pagination-button-disabled"
          data-test="pagination-dots"
          disabled={true}
          type="button"
        >
          ...
        </button>
      </li>
      <li
        className="pagination-item"
        key="2"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          30
        </button>
      </li>
      <li
        className="pagination-item"
        key="3"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          31
        </button>
      </li>
      <li
        className="pagination-item"
        key="4"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          32
        </button>
      </li>
      <li
        className="pagination-item"
        key="5"
      >
        <button
          className="pagination-button pagination-button-active"
          data-test="pagination-button"
          disabled="active"
          onClick={[Function]}
          type="button"
        >
          33
        </button>
      </li>
      <li
        className="pagination-item"
        key="6"
      >
        <button
          className="pagination-button"
          data-test="pagination-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          34
        </button>
      </li>
      <li
        className="pagination-item"
      >
        <button
          className="pagination-button"
          data-test="pagination-chevron-right"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <ForwardRef(SvgChevronRight)
            className="pagination-chevron"
          >
            <svg
              className="pagination-chevron"
            >
              chevron-right.svg
            </svg>
          </ForwardRef(SvgChevronRight)>
        </button>
      </li>
    </ul>
  </nav>
</Pagination>
`;
